From 88f98a4f58e595c77917e38e899dae9602edf5f9 Mon Sep 17 00:00:00 2001
From: Quynh Nguyen <quynh.nguyen.xb@renesas.com>
Date: Thu, 16 Jul 2020 15:16:35 +0700
Subject: [PATCH 245/445] serial: sh-sci: Add lock to prevent the SCSCR
 register access conflict

Commit: <6f96fbf5ba9ac18e1e50e7ebd62b7dc29140fd10>
From:
git://git.kernel.org/pub/scm/linux/kernel/git/horms/renesas-backport.git

sci_start_tx() changes the SCSCR register, but does not lock
sci_port.
Therefore, a caller should lock sci_port.

sci_tx_dma_release() does not lock sci_port in calling
sci_start_tx().
This may cause conflicts of settings to the SCSCR register.
So, this patch adds lock in sci_tx_dma_release().

Signed-off-by: Mitsuru Ezo <mitsuru.ezo.pz@ps.hitachi-solutions.com>
Signed-off-by: Kazuya Mizuguchi <kazuya.mizuguchi.ks@renesas.com>
Signed-off-by: tungvo <tung.vo.eb@rvc.renesas.com>
Signed-off-by: Quynh Nguyen <quynh.nguyen.xb@renesas.com>
---
 drivers/tty/serial/sh-sci.c | 5 ++++-
 1 file changed, 4 insertions(+), 1 deletion(-)

diff --git a/drivers/tty/serial/sh-sci.c b/drivers/tty/serial/sh-sci.c
index d0fb29b..eda52b4 100644
--- a/drivers/tty/serial/sh-sci.c
+++ b/drivers/tty/serial/sh-sci.c
@@ -1187,8 +1187,11 @@ static void sci_tx_dma_release(struct sci_port *s, bool enable_pio)
 		dma_release_channel(chan);
 	}
 
-	if (enable_pio)
+	if (enable_pio) {
+		spin_lock_irqsave(&port->lock, flags);
 		sci_start_tx(port);
+		spin_unlock_irqrestore(&port->lock, flags);
+	}
 }
 
 static void sci_submit_rx(struct sci_port *s)
-- 
2.7.4

